src/Compiler/Microsoft.CodeAnalysis.Razor.Compiler/src/Language/Components/ComponentBindLoweringPass.cs

```diff
@@ -213,120 +239,143 @@ protected override void ExecuteCore(RazorCodeDocument codeDocument, DocumentInte
                     {
-                        duplicateAttributeName = duplicateParameterAttribute.AttributeName;
-                        duplicateTagHelper = duplicateParameterAttribute.TagHelper;
+                        continue;
                     }
-                    if (duplicate != null &&
-                        duplicateTagHelper != null &&
-                        duplicateTagHelper.IsBindTagHelper() &&
-                        duplicateAttributeName == attributeName &&
-                        !object.ReferenceEquals(attribute, duplicate))
+
+                    if (attribute.AttributeName == candidate.AttributeName)
                     {
-                        // Found a duplicate - remove the 'fallback' in favor of the
-                        // more specific tag helper.
-                        node.Children.RemoveAt(i);
+                        // Found a duplicate - remove the 'fallback' in favor of the more specific tag helper.
+                        toRemove.Add(attribute);
                         break;
                     }
                 }
             }
+        }
 
-            // Also treat the general <input @bind="..." /> as a 'fallback' for that case and remove it.
-            // This is a workaround for a limitation where you can't write a tag helper that binds only
-            // when a specific attribute is **not** present.
-            if (attribute != null &&
-                tagHelper != null &&
-                tagHelper.IsInputElementFallbackBindTagHelper())
+        // Now that we've identified attributes that should be removed, iterate them in reverse order
+        // and remove them using the index we stored. Note: We know the attributes are already ordered
+        // by index.
+        for (var i = toRemove.Count - 1; i >= 0; i--)
+        {
+            children.RemoveAt(toRemove[i].Index);
+        }
+
+        // If we still have duplicates at this point then they are genuine conflicts.
+        // Use a hash set to quickly determine whether there are any duplicates.
+        // If so, we need to do a more expensive pass to identify and remove them.
+        using var _ = StringHashSetPool.Ordinal.GetPooledObject(out var duplicates);
+
+        foreach (var child in children)
+        {
+            if (child is TagHelperDirectiveAttributeIntermediateNode { AttributeName: string attributeName } &&
+                !duplicates.Add(attributeName))
             {
-                for (var j = 0; j < node.Children.Count; j++)
+                ReportDiagnosticAndRemoveDuplicates(node);
+                break;
+            }
+        }
+
+        static void ReportDiagnosticAndRemoveDuplicates(IntermediateNode node)
+        {
+            var children = node.Children;
+
+            using var _ = StringDictionaryPool<ImmutableArray<AttributeInfo>.Builder>.Ordinal.GetPooledObject(out var duplicates);
```

Comment:
>AttributeInfo [](http://example.com/codeflow?start=62&length=13)

Why use an AttributeInfo? The builder.Select call only seems to use the Node and not the Index.

Reply:
Never mind, I see now, it's in a later toRemove.Sort call